### PR TITLE
Codereview

### PR DESCRIPTION
**Pull request by kisom**
Let's code review this thang.

### CODE REVIEW

File: filecache.go [hunk 1]
```diff
@@ -47,17 +48,40 @@ var NewCachePipeSize = 4
 
 type cacheItem struct {
 	content    []byte
+	lock       *sync.Mutex
```
Comment (darkhelmet) [inline filecache.go]: Typically you don't need to do a pointer for `sync.Mutex`, then you don't have to worry about making one. The zero value for it is an unlocked mutex.

File: filecache.go [hunk 2]
```diff
@@ -47,17 +48,40 @@ var NewCachePipeSize = 4
...
 	Size       int64
 	Lastaccess time.Time
 	Modified   time.Time
 }
 
+func (itm *cacheItem) Lock() {
+	if itm.lock == nil {
+		return
+	}
```
Comment (darkhelmet) [inline filecache.go]: Then you can get rid of these checks.

File: filecache.go [hunk 3]
```diff
@@ -211,11 +288,13 @@ func (cache *FileCache) changed(name string) bool {
...
 // Expired returns true if the item has not been accessed recently.
 func (cache *FileCache) expired(name string) bool {
-	itm, ok := cache.items[name]
+	itm, ok := cache.getItem(name)
 	if !ok {
 		return true
 	}
+	itm.Lock()
 	dur := time.Now().Sub(itm.Lastaccess)
+	itm.Unlock()
```
Comment (darkhelmet) [inline filecache.go]: `FileCache` shouldn't be locking/unlocking items, they should lock/unlock themselves.

File: filecache.go (hunk 1 above)
Comment (darkhelmet) [inline filecache.go]: You could probably even just embed it, which seems like a perfect use of embedding, but I've never actually done it that way.

Comment (kisom) [inline filecache.go]: Not sure what embedding is here?

**Comment by kisom** [comment]
Most of your comments should now be addressed...

**Comment by darkhelmet** [comment]
Looks good. I'll have to wade through all the locking carefully in the file to reason about it (it's hard in the diff view), but this looks pretty solid.

File: filecache.go [hunk 4]
```diff
@@ -47,21 +48,36 @@ var NewCachePipeSize = 4
 
 type cacheItem struct {
 	content    []byte
+	lock       sync.Mutex
```
Comment (darkhelmet) [inline filecache.go:51]: For embedding you could just do

``` go
type cacheItem struct {
  sync.Mutex
  content []byte
  ...
}
```

but then you'd have the `Lock` and `Unlock` methods on the `cacheItem` type, so probably not a great idea.

**Comment by kisom** [comment]
I ran it with the tests with the race detector and it doesn't complain anymore.

**Comment by kisom** [comment]
I also handwrote a call graph to figure out which parts needed to be locked :)

**Comment by darkhelmet** [comment]
Sweet! Was the race detector complaining prior to the refactor?
—

**Comment by kisom** [comment]
It gave me 17 pages of output -- I printed that and the source and started correlating where the issues were occurring; that was the starting point for figuring out where things needed to be locked.

**Comment by darkhelmet** [comment]
Can you throw the 17 pages in a gist? I would be interested in seeing it. 
—

**Comment by kisom** [comment]
Here you are: https://gist.github.com/kisom/f992b76fbb97218e902d

TL;DR: Found 42 data race(s)

**Comment by kisom** [comment]
I went ahead and merged this into master.
